### PR TITLE
add context to Fetch* methods

### PR DESCRIPTION
**Pull request by JekaMas**
as a developer I need an API that provides me cancellation

### CODE REVIEW

File: consensus/bor/heimdall/client.go [hunk 1]
```diff
@@ -126,58 +131,69 @@ func (h *HeimdallClient) FetchLatestCheckpoint() (*checkpoint.Checkpoint, error)
...
+
+	body, err := internalFetchWithTimeout(ctx, client, url)
+
+	if body == nil {
+		return nil, ErrNoResponse
+	}
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (manav2401) [inline consensus/bor/heimdall/client.go:187]: Should we replace the above code blocks with each other? Isn't there a possibility of having a nil body with some error and if yes, the we won't be able to send the error to the parent function.

File: consensus/bor/heimdall/client.go [hunk 2]
```diff
@@ -126,58 +131,69 @@ func (h *HeimdallClient) FetchLatestCheckpoint() (*checkpoint.Checkpoint, error)
...
-			cancel()
-
-			if err == nil && body != nil {
-				err = json.Unmarshal(body, result)
-				if err != nil {
-					return nil, err
-				}
+			result, err = Fetch[T](ctx, client, url)
 
+			// ignore or log the error
```
Comment (manav2401) [inline consensus/bor/heimdall/client.go:167]: Can we do something like logging error only if attempt > N?

File: eth/backend.go [hunk 3]
```diff
@@ -632,7 +633,8 @@ func (s *Ethereum) startCheckpointWhitelistService() {
 	}
 
 	// first run the checkpoint whitelist
-	err := s.handleWhitelistCheckpoint()
+	// TODO: add context timeout if needed
```
[the remark this answers is not in the document]
Comment (JekaMas) [reply eth/backend.go:636]: yeap, it is for RFC-35 label

File: eth/backend.go [hunk 4]
```diff
@@ -647,7 +649,8 @@ func (s *Ethereum) startCheckpointWhitelistService() {
 	for {
 		select {
 		case <-ticker.C:
-			err := s.handleWhitelistCheckpoint()
+			// TODO: add context timeout if needed
```
[the remark this answers is not in the document]
Comment (JekaMas) [reply eth/backend.go:652]: yes, it is for RFC35

File: consensus/bor/heimdall/client.go [hunk 5]
```diff
@@ -126,58 +131,69 @@ func (h *HeimdallClient) FetchLatestCheckpoint() (*checkpoint.Checkpoint, error)
...
+	attempt := 1
+
 	for {
 		log.Info("Retrying again in 5 seconds to fetch data from Heimdall", "path", url.Path, "attempt", attempt)
+
 		attempt++
+
 		select {
+		case <-ctx.Done():
+			log.Debug("Shutdown detected, terminating request")
```
[the remark this answers is not in the document]
Comment (JekaMas) [reply consensus/bor/heimdall/client.go]: will do.

File: consensus/bor/heimdall/client.go [hunk 6]
```diff
@@ -126,58 +131,69 @@ func (h *HeimdallClient) FetchLatestCheckpoint() (*checkpoint.Checkpoint, error)
...
 	// create a new ticker for retrying the request
-	ticker := time.NewTicker(5 * time.Second)
+	ticker := time.NewTicker(retryCall)
 	defer ticker.Stop()
 
+	// attempt counter
+	attempt := 1
+
 	for {
 		log.Info("Retrying again in 5 seconds to fetch data from Heimdall", "path", url.Path, "attempt", attempt)
```
[the remark this answers is not in the document]
Comment (manav2401) [reply consensus/bor/heimdall/client.go:154]: I meant in the log statement but it's not necessary.

[the remark this answers is not in the document]
Comment (JekaMas) [reply consensus/bor/heimdall/client.go:154]: I log the attempt number instead

**Review by manav2401 — APPROVED** [review]

**Review by 0xKrishna — APPROVED** [review]

**Pull request merged**